srv/salt/_modules/packagemanager.py

```diff
@@ -0,0 +1,275 @@
+            self.pm = Apt(**kwargs)
+        else:
+            raise ValueError("Failed to detect PackageManager for OS."
+                             "Open an issue on github.com/SUSE/DeepSea")
+
+    def _reboot(self):
+        """
+        Assuming `shutdown -r` works on all platforms
+        """
+        log.info("The PackageManager asked for a system reboot. Rebooting in 1 Minute")
+        if self.debug or not self.reboot:
+            log.debug("Faking Reboot")
+            return None
+        log.debug("Initializing Reboot.")
+        cmd = "shutdown -r"
+        Popen(cmd, stdout=PIPE, shell=True)
+
+
+class Apt(PackageManager):
+
+    VERSION = 0.1
+
+    def __init__(self, **kwargs):
+        """
+        Instead of reinitializing __init__ from super,
+        the child get kwargs passed as arguments to
+        have the possibility to handle those differently.
+        """
+        self.kernel = kwargs.get('kernel', False)
+        self.debug = kwargs.get('debug', False)
+        self.reboot = kwargs.get('reboot', True)
+        self.base_flags = ['--yes']
+
+    def _updates_needed(self):
+        """
+        Checking the output of apt-check for
+        (1) Regular updates
+        (2) Security updates
+        Content is written to stderr
+        """
+        self._refresh()
+        cmd = "/usr/lib/update-notifier/apt-check"
+        proc = Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True)
+        stdout, stderr = proc.communicate()
+        for cn in stderr.split(";"):
+            if int(cn) > 0:
+                log.info('Update Needed')
+                return True
+            log.info('No Update Needed')
+        return False
+
+    def _refresh(self):
+        """
+        Resynchronize the package index files from their sources
+        """
+        cmd = 'apt-get {} update'.format(self.base_flags)
+        Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True)
+
+    def _handle(self, strat='up'):
+        """
+        Conbines up and dup and executes the constructed zypper command.
+        """
+        if strat == 'up':
+            strat = 'upgrade'
+        if self._updates_needed():
+            base_command = ['apt-get']
+            strategy_flags = ["-o Dpkg::Options::=",
+                              "--allow-change-held-packages", "-fuy"]
+            if self.debug:
+                strategy_flags.append("--dry-run")
+            base_command.extend(self.base_flags)
+            base_command.extend(strat.split())
+            base_command.extend(strategy_flags)
+            proc = Popen(base_command, stdout=PIPE, stderr=PIPE)
+            stdout, stderr = proc.communicate()
+            for line in stdout:
+                log.info(line)
+            for line in stderr:
+                log.info(line)
+            log.info("returncode: {}".format(proc.returncode))
+            if proc.returncode == 0:
+                if os.path.isfile('/var/run/reboot-required'):
+                    self._reboot()
```

Comment:
I added `autoreboot: False|True` to globally set the default behaviour on whether deepsea has the right to 'automagically reboot' in case the system requests it. So see it as a 'user variable' rather than an internal one.

My intention was to do it the way around. I mentioned in another comment that changing to 'zypper patch' will get rid of the ceph/update/reboot sls file. 

and so technically double reboots won't be possible in zypper right now, as the 107 will never be received (due to zypper _up_'s missing ret implementation). They may be with Apt but as `Packagemanager` will cause a reboot, the next step in the orchestrate file (after the reboot) is ceph.upgrade.reboot which compares the 'installed' vs 'used' kernel, which is the same after the reboot.